### PR TITLE
[Mliestone-3] changes in OrderedMap

### PR DESCRIPTION
**Pull request by GoPavel**
# Summary: 

+ add `any`/`all` functions
+ add `contains` function
+ add `minEntry`/`maxEntry`
+ Store `size` in the Map 
+ add `entriesRev()`, remove `iter()`

# Benchmark results

| |binary_size|generate|max mem|batch_get 50|batch_put 50|batch_remove 50|upgrade|
|--:|--:|--:|--:|--:|--:|--:|--:|
|persistentmap_baseline_100|192_546|204_118|44_108|[51_387](persistentmap_baseline_100_get.svg)|[124_116](persistentmap_baseline_100_put.svg)|[128_293](persistentmap_baseline_100_remove.svg)|[388_678](persistentmap_baseline_100_upgrade.svg)|
|persistentmap_100|198_522|207_689|44_180|[52_994](persistentmap_100_get.svg)|[133_418](persistentmap_100_put.svg)|[137_455](persistentmap_100_remove.svg)|[391_376](persistentmap_100_upgrade.svg)|
|rbtree_100|190_190|225_155|42_540|[50_045](rbtree_100_get.svg)|[135_367](rbtree_100_put.svg)|[133_686](rbtree_100_remove.svg)|[566_251](rbtree_100_upgrade.svg)|
|persistentmap_baseline_1000|192_546|2_775_808|596_916|[68_754](persistentmap_baseline_1000_get.svg)|[162_897](persistentmap_baseline_1000_put.svg)|[173_291](persistentmap_baseline_1000_remove.svg)|[3_635_553](persistentmap_baseline_1000_upgrade.svg)|
|persistentmap_1000|198_522|2_810_879|596_988|[70_361](persistentmap_1000_get.svg)|[171_804](persistentmap_1000_put.svg)|[182_848](persistentmap_1000_remove.svg)|[3_637_898](persistentmap_1000_upgrade.svg)|
|rbtree_1000|190_190|3_118_490|580_948|[67_516](rbtree_1000_get.svg)|[181_375](rbtree_1000_put.svg)|[180_396](rbtree_1000_remove.svg)|[5_410_399](rbtree_1000_upgrade.svg)|
|persistentmap_baseline_10000|192_546|42_949_367|360_396|[85_219](persistentmap_baseline_10000_get.svg)|[199_435](persistentmap_baseline_10000_put.svg)|[221_201](persistentmap_baseline_10000_remove.svg)|[36_039_486](persistentmap_baseline_10000_upgrade.svg)|
|persistentmap_10000|198_522|43_300_700|360_468|[86_826](persistentmap_10000_get.svg)|[208_342](persistentmap_10000_put.svg)|[230_758](persistentmap_10000_remove.svg)|[36_042_144](persistentmap_10000_upgrade.svg)|
|rbtree_10000|190_190|51_301_049|520_428|[83_465](rbtree_10000_get.svg)|[224_135](rbtree_10000_put.svg)|[230_257](rbtree_10000_remove.svg)|[53_867_183](rbtree_10000_upgrade.svg)|
|persistentmap_baseline_100000|192_546|509_484_639|3_600_396|[99_728](persistentmap_baseline_100000_get.svg)|[238_691](persistentmap_baseline_100000_put.svg)|[265_873](persistentmap_baseline_100000_remove.svg)|[458_331_809](persistentmap_baseline_100000_upgrade.svg)|
|persistentmap_100000|198_522|512_985_316|3_600_468|[101_335](persistentmap_100000_get.svg)|[247_557](persistentmap_100000_put.svg)|[275_430](persistentmap_100000_remove.svg)|[458_336_050](persistentmap_100000_upgrade.svg)|
|rbtree_100000|190_190|606_914_881|5_200_428|[98_012](rbtree_100000_get.svg)|[270_465](rbtree_100000_put.svg)|[276_802](rbtree_100000_remove.svg)|[698_929_134](rbtree_100000_upgrade.svg)|
|persistentmap_baseline_1000000|192_546|5_883_267_983|36_000_432|[118_499](persistentmap_baseline_1000000_get.svg)|[278_670](persistentmap_baseline_1000000_put.svg)|[317_512](persistentmap_baseline_1000000_remove.svg)|[4_583_349_359](persistentmap_baseline_1000000_upgrade.svg)|
|persistentmap_1000000|198_522|5_918_268_045|36_000_504|[120_106](persistentmap_1000000_get.svg)|[287_536](persistentmap_1000000_put.svg)|[327_069](persistentmap_1000000_remove.svg)|[4_583_354_092](persistentmap_1000000_upgrade.svg)|
|rbtree_1000000|190_190|6_993_676_959|52_000_464|[116_417](rbtree_1000000_get.svg)|[317_299](rbtree_1000000_put.svg)|[330_296](rbtree_1000000_remove.svg)|[6_988_883_792](rbtree_1000000_upgrade.svg)|

## Persistent map API

| |size|foldLeft|foldRight|mapfilter|map|
|--:|--:|--:|--:|--:|--:|
|persistentmap|100|[19_647](persistentmap_foldLeft_100.svg)|[19_559](persistentmap_foldRight_100.svg)|[91_897](persistentmap_mapfilter_100.svg)|[26_359](persistentmap_map_100.svg)|
|persistentmap_baseline|100|[19_623](persistentmap_baseline_foldLeft_100.svg)|[19_535](persistentmap_baseline_foldRight_100.svg)|[89_915](persistentmap_baseline_mapfilter_100.svg)|[26_278](persistentmap_baseline_map_100.svg)|
|persistentmap|1000|[165_657](persistentmap_foldLeft_1000.svg)|[164_169](persistentmap_foldRight_1000.svg)|[1_599_215](persistentmap_mapfilter_1000.svg)|[230_405](persistentmap_map_1000.svg)|
|persistentmap_baseline|1000|[165_633](persistentmap_baseline_foldLeft_1000.svg)|[164_145](persistentmap_baseline_foldRight_1000.svg)|[1_577_927](persistentmap_baseline_mapfilter_1000.svg)|[230_324](persistentmap_baseline_map_1000.svg)|
|persistentmap|10000|[1_628_063](persistentmap_foldLeft_10000.svg)|[1_609_832](persistentmap_foldRight_10000.svg)|[29_896_780](persistentmap_mapfilter_10000.svg)|[2_270_579](persistentmap_map_10000.svg)|
|persistentmap_baseline|10000|[1_628_080](persistentmap_baseline_foldLeft_10000.svg)|[1_609_767](persistentmap_baseline_foldRight_10000.svg)|[29_675_517](persistentmap_baseline_mapfilter_10000.svg)|[2_270_416](persistentmap_baseline_map_10000.svg)|
|persistentmap|100000|[16_254_154](persistentmap_foldLeft_100000.svg)|[16_059_713](persistentmap_foldRight_100000.svg)|[361_440_658](persistentmap_mapfilter_100000.svg)|[94_266_689](persistentmap_map_100000.svg)|
|persistentmap_baseline|100000|[16_254_171](persistentmap_baseline_foldLeft_100000.svg)|[16_059_689](persistentmap_baseline_foldRight_100000.svg)|[359_232_494](persistentmap_baseline_mapfilter_100000.svg)|[94_265_059](persistentmap_baseline_map_100000.svg)|
|persistentmap|1000000|[162_559_245](persistentmap_foldLeft_1000000.svg)|[876_571_882](persistentmap_foldRight_1000000.svg)|[4_220_901_740](persistentmap_mapfilter_1000000.svg)|[942_590_783](persistentmap_map_1000000.svg)|
|persistentmap_baseline|1000000|[162_559_221](persistentmap_baseline_foldLeft_1000000.svg)|[876_570_432](persistentmap_baseline_foldRight_1000000.svg)|[4_198_876_914](persistentmap_baseline_mapfilter_1000000.svg)|[942_589_235](persistentmap_baseline_map_1000000.svg)|

### CODE REVIEW

File: src/PersistentOrderedMap.mo [hunk 1]
```diff
@@ -527,6 +527,50 @@ module {
...
+    /// Example:
+    /// ```
+    /// import Map "mo:base/PersistentOrderedMap";
+    /// import Nat "mo:base/Nat";
+    /// import Debug "mo:base/Debug";
+    ///
+    /// let mapOps = Map.MapOps<Nat>(Nat.compare);
+    /// let map = mapOps.fromIter<Text>(Iter.fromArray([(0, "0"), (2, "2"), (1, "1")]));
+    ///
+    /// Debug.print(debug_show(mapOps.all(map, func (k, v) = (v == debug_show(k)))));
```
Comment (s-and-witch) [inline src/PersistentOrderedMap.mo:590]: Can you give negative example here and positive for `some`? BTW, I didn't know that `=`-syntax works for lambdas.

File: src/PersistentOrderedMap.mo [hunk 2]
```diff
@@ -227,6 +227,50 @@ module {
...
+    /// let map = mapOps.fromIter<Text>(Iter.fromArray([(0, "Zero"), (2, "Two"), (1, "One")]));
+    ///
+    /// Debug.print(debug_show(mapOps.maxEntry(map))); // => ?(2, "Two")
+    /// Debug.print(debug_show(mapOps.maxEntry(mapOps.empty()))); // => null
+    /// ```
+    ///
+    /// Runtime: `O(log(n))`
+    /// Space: `O(1)`.
+    /// where `n` denotes the number of key-value entries stored in the map and
+    /// assuming that the `compare` function implements an `O(1)` comparison.
```
Comment (s-and-witch) [inline src/PersistentOrderedMap.mo]: I think it doesn't require this assumption? We don't call `compare` here, just unconditionally take leftmost/rightmost element.

File: src/PersistentOrderedMap.mo [hunk 3]
```diff
@@ -227,6 +227,50 @@ module {
...
+    /// let map = mapOps.fromIter<Text>(Iter.fromArray([(0, "Zero"), (2, "Two"), (1, "One")]));
+    ///
+    /// Debug.print(debug_show(mapOps.minEntry(map))); // => ?(0, "Zero")
+    /// Debug.print(debug_show(mapOps.minEntry(mapOps.empty()))); // => null
+    /// ```
+    ///
+    /// Runtime: `O(log(n))`
+    /// Space: `O(1)`.
+    /// where `n` denotes the number of key-value entries stored in the map and
+    /// assuming that the `compare` function implements an `O(1)` comparison.
```
Comment (s-and-witch) [inline src/PersistentOrderedMap.mo]: The same as above applies here.

File: src/PersistentOrderedMap.mo [hunk 4]
```diff
@@ -754,6 +798,36 @@ module {
       }
     };
 
+    public func maxEntry<K, V>(m: Map<K, V>): ?(K, V) {
+      func rightmost(m: Map<K, V>): ?(K, V) {
```
Comment (s-and-witch) [inline src/PersistentOrderedMap.mo]: We can return `(K, V)` here and create `?rightmost(m)` on the outer call. I'm not sure about performance implications, but in haskell I would say that such code would be more optimisations friendly.

File: test/PersistentOrderedMap.test.mo [hunk 5]
```diff
@@ -177,6 +177,11 @@ run(
         natMapOps.contains(buildTestMap(), 0),
         M.equals(T.bool(false))
       ),
+      test(
+        "maxEntry",
```
Comment (s-and-witch) [inline test/PersistentOrderedMap.test.mo:176]: Please, add tests for `minEntry` (here and below)

**Review by s-and-witch — APPROVED** [review]
Overall looks good, just several suggestions.

File: src/PersistentOrderedMap.mo (hunk 1 above)
Comment (GoPavel) [reply src/PersistentOrderedMap.mo:590]: It works, when you have one expression. So you have two variants:
* `func x = x`
* func x { x }`

File: test/PersistentOrderedMap.test.mo (hunk 5 above)
Comment (GoPavel) [reply test/PersistentOrderedMap.test.mo:176]: fixed

File: src/PersistentOrderedMap.mo (hunk 4 above)
Comment (GoPavel) [reply src/PersistentOrderedMap.mo]: fixed

File: src/PersistentOrderedMap.mo (hunk 2 above)
Comment (GoPavel) [reply src/PersistentOrderedMap.mo]: fixed

File: src/PersistentOrderedMap.mo (hunk 3 above)
Comment (GoPavel) [reply src/PersistentOrderedMap.mo]: fixed

File: src/PersistentOrderedMap.mo (hunk 1 above)
Comment (GoPavel) [reply src/PersistentOrderedMap.mo:590]: fixed

Comment (GoPavel) [reply src/PersistentOrderedMap.mo:590]: About the syntax: you can check tests for `all` for example. There I also used the same style

**Comment by GoPavel** [comment]
Merged manually into `milestone-3`, see https://github.com/dfinity/motoko-base/pull/662

**Pull request closed without merging**